File: tests/algorithms/x/Kotlin/other/lru_cache.kt
import java.math.BigInteger

fun <T> _listSet(lst: MutableList<T>, idx: Int, v: T) { while (lst.size <= idx) lst.add(v); lst[idx] = v }

data class Node(var key: Int = 0, var value: Int = 0, var prev: Int = 0, var next: Int = 0)
data class DoubleLinkedList(var nodes: MutableList<Node> = mutableListOf<Node>(), var head: Int = 0, var tail: Int = 0)
data class LRUCache(var list: DoubleLinkedList = DoubleLinkedList(nodes = mutableListOf<Node>(), head = 0, tail = 0), var capacity: Int = 0, var num_keys: Int = 0, var hits: Int = 0, var misses: Int = 0, var cache: MutableMap<String, Int> = mutableMapOf<String, Int>())
data class GetResult(var cache: LRUCache = LRUCache(list = DoubleLinkedList(nodes = mutableListOf<Node>(), head = 0, tail = 0), capacity = 0, num_keys = 0, hits = 0, misses = 0, cache = mutableMapOf<String, Int>()), var value: Int = 0, var ok: Boolean = false)
fun new_list(): DoubleLinkedList {
    var nodes: MutableList<Node> = mutableListOf<Node>()
    var head: Node = Node(key = 0, value = 0, prev = 0 - 1, next = 1)
    var tail: Node = Node(key = 0, value = 0, prev = 0, next = 0 - 1)
    nodes = run { val _tmp = nodes.toMutableList(); _tmp.add(head); _tmp }
    nodes = run { val _tmp = nodes.toMutableList(); _tmp.add(tail); _tmp }
    return DoubleLinkedList(nodes = nodes, head = 0, tail = 1)
}

fun dll_add(lst: DoubleLinkedList, idx: Int): DoubleLinkedList {
    var nodes: MutableList<Node> = lst.nodes
    var tail_idx: Int = (lst.tail).toInt()
    var tail_node: Node = nodes[tail_idx]!!
    var prev_idx: Int = (tail_node.prev).toInt()
    var node: Node = nodes[idx]!!
    node.prev = prev_idx
    node.next = tail_idx
    _listSet(nodes, idx, node)
    var prev_node: Node = nodes[prev_idx]!!
    prev_node.next = idx
    _listSet(nodes, prev_idx, prev_node)
    tail_node.prev = idx
    _listSet(nodes, tail_idx, tail_node)
    lst.nodes = nodes
    return lst
}

fun dll_remove(lst: DoubleLinkedList, idx: Int): DoubleLinkedList {
    var nodes: MutableList<Node> = lst.nodes
    var node: Node = nodes[idx]!!
    var prev_idx: Int = (node.prev).toInt()
    var next_idx: Int = (node.next).toInt()
    if ((prev_idx == (0 - 1)) || (next_idx == (0 - 1))) {
        return lst
    }
    var prev_node: Node = nodes[prev_idx]!!
    prev_node.next = next_idx
    _listSet(nodes, prev_idx, prev_node)
    var next_node: Node = nodes[next_idx]!!
    next_node.prev = prev_idx
    _listSet(nodes, next_idx, next_node)
    node.prev = 0 - 1
    node.next = 0 - 1
    _listSet(nodes, idx, node)
    lst.nodes = nodes
    return lst
}

fun new_cache(cap: Int): LRUCache {
    var empty_map: MutableMap<String, Int> = mutableMapOf<String, Int>()
    return LRUCache(list = new_list(), capacity = cap, num_keys = 0, hits = 0, misses = 0, cache = empty_map)
}

fun lru_get(c: LRUCache, key: Int): GetResult {
    var cache: LRUCache = c
    var key_str: String = key.toString()
    if (key_str in cache.cache) {
        var idx: Int = ((cache.cache)[key_str] as Int).toInt()
        if (idx != (0 - 1)) {
            cache.hits = cache.hits + 1
            var node: Node = (cache.list.nodes)[idx]!!
            var value: Int = (node.value).toInt()
            cache.list = dll_remove(cache.list, idx)
            cache.list = dll_add(cache.list, idx)
            return GetResult(cache = cache, value = value, ok = true)
        }
    }
    cache.misses = cache.misses + 1
    return GetResult(cache = cache, value = 0, ok = false)
}

fun lru_put(c: LRUCache, key: Int, value: Int): LRUCache {
    var cache: LRUCache = c
    var key_str: String = key.toString()
    if (!(key_str in cache.cache)) {
        if (cache.num_keys >= cache.capacity) {
            var head_node: Node = (cache.list.nodes)[cache.list.head]!!
            var first_idx: Int = (head_node.next).toInt()
            var first_node: Node = (cache.list.nodes)[first_idx]!!
            var old_key: Int = (first_node.key).toInt()
            cache.list = dll_remove(cache.list, first_idx)
            var mdel: MutableMap<String, Int> = cache.cache
            (mdel)[old_key.toString()] = 0 - 1
            cache.cache = mdel
            cache.num_keys = cache.num_keys - 1
        }
        var nodes: MutableList<Node> = cache.list.nodes
        var new_node: Node = Node(key = key, value = value, prev = 0 - 1, next = 0 - 1)
        nodes = run { val _tmp = nodes.toMutableList(); _tmp.add(new_node); _tmp }
        var idx: Int = (nodes.size - 1).toInt()
        cache.list.nodes = nodes
        cache.list = dll_add(cache.list, idx)
        var m: MutableMap<String, Int> = cache.cache
        (m)[key_str] = idx
        cache.cache = m
        cache.num_keys = cache.num_keys + 1
    } else {
        var m: MutableMap<String, Int> = cache.cache
        var idx: Int = ((m)[key_str] as Int).toInt()
        var nodes: MutableList<Node> = cache.list.nodes
        var node: Node = nodes[idx]!!
        node.value = value
        _listSet(nodes, idx, node)
        cache.list.nodes = nodes
        cache.list = dll_remove(cache.list, idx)
        cache.list = dll_add(cache.list, idx)
        cache.cache = m
    }
    return cache
}

fun cache_info(cache: LRUCache): String {
    return ((((((("CacheInfo(hits=" + cache.hits.toString()) + ", misses=") + cache.misses.toString()) + ", capacity=") + cache.capacity.toString()) + ", current size=") + cache.num_keys.toString()) + ")"
}

fun print_result(res: GetResult): Unit {
    if (((res.ok) as Boolean)) {
        println(res.value.toString())
    } else {
        println("None")
    }
}

fun user_main(): Unit {
    var cache: LRUCache = new_cache(2)
    cache = lru_put(cache, 1, 1)
    cache = lru_put(cache, 2, 2)
    var r1: GetResult = lru_get(cache, 1)
    cache = r1.cache
    print_result(r1)
    cache = lru_put(cache, 3, 3)
    var r2: GetResult = lru_get(cache, 2)
    cache = r2.cache
    print_result(r2)
    cache = lru_put(cache, 4, 4)
    var r3: GetResult = lru_get(cache, 1)
    cache = r3.cache
    print_result(r3)
    var r4: GetResult = lru_get(cache, 3)
    cache = r4.cache
    print_result(r4)
    var r5: GetResult = lru_get(cache, 4)
    cache = r5.cache
    print_result(r5)
    println(cache_info(cache))
}

fun main() {
    user_main()
}

File: tests/algorithms/x/PHP/other/lru_cache.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function new_list() {
  $nodes = [];
  $head = ['key' => 0, 'next' => 1, 'prev' => 0 - 1, 'value' => 0];
  $tail = ['key' => 0, 'next' => 0 - 1, 'prev' => 0, 'value' => 0];
  $nodes = _append($nodes, $head);
  $nodes = _append($nodes, $tail);
  return ['head' => 0, 'nodes' => $nodes, 'tail' => 1];
};
  function dll_add($lst, $idx) {
  $nodes = $lst['nodes'];
  $tail_idx = $lst['tail'];
  $tail_node = $nodes[$tail_idx];
  $prev_idx = $tail_node['prev'];
  $node = $nodes[$idx];
  $node['prev'] = $prev_idx;
  $node['next'] = $tail_idx;
  $nodes[$idx] = $node;
  $prev_node = $nodes[$prev_idx];
  $prev_node['next'] = $idx;
  $nodes[$prev_idx] = $prev_node;
  $tail_node['prev'] = $idx;
  $nodes[$tail_idx] = $tail_node;
  $lst['nodes'] = $nodes;
  return $lst;
};
  function dll_remove($lst, $idx) {
  $nodes = $lst['nodes'];
  $node = $nodes[$idx];
  $prev_idx = $node['prev'];
  $next_idx = $node['next'];
  if ($prev_idx == 0 - 1 || $next_idx == 0 - 1) {
  return $lst;
}
  $prev_node = $nodes[$prev_idx];
  $prev_node['next'] = $next_idx;
  $nodes[$prev_idx] = $prev_node;
  $next_node = $nodes[$next_idx];
  $next_node['prev'] = $prev_idx;
  $nodes[$next_idx] = $next_node;
  $node['prev'] = 0 - 1;
  $node['next'] = 0 - 1;
  $nodes[$idx] = $node;
  $lst['nodes'] = $nodes;
  return $lst;
};
  function new_cache($cap) {
  $empty_map = [];
  return ['cache' => $empty_map, 'capacity' => $cap, 'hits' => 0, 'list' => new_list(), 'misses' => 0, 'num_keys' => 0];
};
  function lru_get($c, $key) {
  $cache = $c;
  $key_str = _str($key);
  if (isset($cache['cache'][$key_str])) {
  $idx = $cache['cache'][$key_str];
  if ($idx != 0 - 1) {
  $cache['hits'] = $cache['hits'] + 1;
  $node = $cache['list']['nodes'][$idx];
  $value = $node['value'];
  $cache['list'] = dll_remove($cache['list'], $idx);
  $cache['list'] = dll_add($cache['list'], $idx);
  return ['cache' => $cache, 'ok' => true, 'value' => $value];
};
}
  $cache['misses'] = $cache['misses'] + 1;
  return ['cache' => $cache, 'ok' => false, 'value' => 0];
};
  function lru_put($c, $key, $value) {
  $cache = $c;
  $key_str = _str($key);
  if (!(isset($cache['cache'][$key_str]))) {
  if ($cache['num_keys'] >= $cache['capacity']) {
  $head_node = $cache['list']['nodes'][$cache['list']['head']];
  $first_idx = $head_node['next'];
  $first_node = $cache['list']['nodes'][$first_idx];
  $old_key = $first_node['key'];
  $cache['list'] = dll_remove($cache['list'], $first_idx);
  $mdel = $cache['cache'];
  $mdel[_str($old_key)] = 0 - 1;
  $cache['cache'] = $mdel;
  $cache['num_keys'] = $cache['num_keys'] - 1;
};
  $nodes = $cache['list']['nodes'];
  $new_node = ['key' => $key, 'next' => 0 - 1, 'prev' => 0 - 1, 'value' => $value];
  $nodes = _append($nodes, $new_node);
  $idx = count($nodes) - 1;
  $cache['list']['nodes'] = $nodes;
  $cache['list'] = dll_add($cache['list'], $idx);
  $m = $cache['cache'];
  $m[$key_str] = $idx;
  $cache['cache'] = $m;
  $cache['num_keys'] = $cache['num_keys'] + 1;
} else {
  $m = $cache['cache'];
  $idx = $m[$key_str];
  $nodes = $cache['list']['nodes'];
  $node = $nodes[$idx];
  $node['value'] = $value;
  $nodes[$idx] = $node;
  $cache['list']['nodes'] = $nodes;
  $cache['list'] = dll_remove($cache['list'], $idx);
  $cache['list'] = dll_add($cache['list'], $idx);
  $cache['cache'] = $m;
}
  return $cache;
};
  function cache_info($cache) {
  return 'CacheInfo(hits=' . _str($cache['hits']) . ', misses=' . _str($cache['misses']) . ', capacity=' . _str($cache['capacity']) . ', current size=' . _str($cache['num_keys']) . ')';
};
  function print_result($res) {
  if ($res['ok']) {
  echo rtrim(_str($res['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
};
  function main() {
  $cache = new_cache(2);
  $cache = lru_put($cache, 1, 1);
  $cache = lru_put($cache, 2, 2);
  $r1 = lru_get($cache, 1);
  $cache = $r1['cache'];
  print_result($r1);
  $cache = lru_put($cache, 3, 3);
  $r2 = lru_get($cache, 2);
  $cache = $r2['cache'];
  print_result($r2);
  $cache = lru_put($cache, 4, 4);
  $r3 = lru_get($cache, 1);
  $cache = $r3['cache'];
  print_result($r3);
  $r4 = lru_get($cache, 3);
  $cache = $r4['cache'];
  print_result($r4);
  $r5 = lru_get($cache, 4);
  $cache = $r5['cache'];
  print_result($r5);
  echo rtrim(cache_info($cache)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
